web-bundle/src/main/java/com/graphhopper/swl/CustomGraphHopperOSM.java

```diff
@@ -0,0 +1,146 @@
+package com.graphhopper.swl;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.graphhopper.reader.DataReader;
+import com.graphhopper.reader.ReaderElement;
+import com.graphhopper.reader.ReaderWay;
+import com.graphhopper.reader.osm.GraphHopperOSM;
+import com.graphhopper.reader.osm.OSMInput;
+import com.graphhopper.reader.osm.OSMReader;
+import com.graphhopper.routing.util.EncodingManager;
+import com.graphhopper.storage.GraphHopperStorage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Custom implementation of internal class GraphHopper uses to parse OSM files into GH's internal graph data structures.
+ * In particular, the purpose of this class is to parse and store specific OSM tag information needed to replicate the
+ * logic used to create the `lanes` and `flags` columns of R5's street network CSV export.
+ *
+ * If needed, this class can easily be extended in the future to parse other OSM tag information for use in exporting
+ * data about a particular region's GH street network.
+ */
+
+public class CustomGraphHopperOSM extends GraphHopperOSM {
+    private static final Logger LOG = LoggerFactory.getLogger(CustomGraphHopperOSM.class);
+
+    // Tags considered by R5 when calculating the value of the `lanes` column
+    private static final Set<String> LANE_TAGS = Sets.newHashSet("lanes", "lanes:forward", "lanes:backward");
+    private String osmPath;
+
+    // Map of OSM way ID -> (Map of OSM lane tag name -> tag value)
+    private Map<Long, Map<String, String>> osmIdToLaneTags;
+    // Map of GH edge ID to OSM way ID
+    private Map<Integer, Long> ghIdToOsmId;
+    // Map of OSM way ID to String[] of access flags (from set {ALLOWS_CAR, ALLOWS_BIKE, ALLOWS_PEDESTRIAN}) for each
+    // edge direction, in order [forward, backward]
+    private Map<Long, String[]> osmIdToAccessFlags;
+
+    public CustomGraphHopperOSM(String osmPath) {
+        super();
+        this.osmPath = osmPath;
+        this.osmIdToLaneTags = Maps.newHashMap();
+        this.ghIdToOsmId = Maps.newHashMap();
+        this.osmIdToAccessFlags = Maps.newHashMap();
+    }
+
+    @Override
+    protected void registerCustomEncodedValues(EncodingManager.Builder emBuilder) {
+        StableIdEncodedValues.createAndAddEncodedValues(emBuilder);
+    }
+
+    /**
+     * Override creation of OSM reader to read the file once at initialization time, for the sole purpose of storing
+     * OSM information that will be used later in the export script.
+     *
+     * Note that this approach requires reading the OSM file twice: once during the static initialization code, and
```

Comment:
@michaz this is awfully hacky 😅 Can you think of any way we could avoid the double-reading of OSM here? (once in the static initialization code, and once in the call to `importOrLoad()` in `ExportCommand`, which is when the overridden `storeOsmWayID` function below gets called)